front/app/modules/commercial/representativeness/admin/hooks/useReferenceData.ts

```diff
@@ -0,0 +1,134 @@
+import { useState, useEffect } from 'react';
+
+// services
+import {
+  usersByRegFieldStream,
+  usersByGenderStream,
+  usersByDomicileStream,
+  TStreamResponse,
+} from 'modules/commercial/user_custom_fields/services/stats';
+
+// utils
+import { isNilOrError, NilOrError } from 'utils/helperUtils';
+
+// typings
+import {
+  IUserCustomFieldData,
+  TCustomFieldCode,
+} from 'modules/commercial/user_custom_fields/services/userCustomFields';
+import { Multiloc } from 'typings';
+
+export interface RepresentativenessRow {
+  name: string;
+  key: string;
+  title_multiloc: Multiloc;
+  actualPercentage: number;
+  referencePercentage: number;
+  actualNumber: number;
+  referenceNumber: number;
+}
+
+export type RepresentativenessData = RepresentativenessRow[];
+
+const getStream = (code: TCustomFieldCode | null) => {
+  if (code === null) return usersByRegFieldStream;
+  if (code === 'gender') return usersByGenderStream;
+  if (code === 'domicile') return usersByDomicileStream;
+  return undefined;
+};
+
+function useReferenceData(field: IUserCustomFieldData, projectId?: string) {
+  const [referenceData, setReferenceData] = useState<
+    RepresentativenessData | NilOrError
+  >();
+  const [includedUserPercentage, setIncludedUserPercentage] = useState<
+    number | NilOrError
+  >();
+  const [referenceDataUploaded, setReferenceDataUploaded] = useState<
+    boolean | undefined
+  >();
+
+  const code = field.attributes.code;
+
+  useEffect(() => {
+    const stream = getStream(code);
+    if (!stream) {
+      setReferenceData(null);
+      return;
+    }
+
+    const observable = stream(
+      { queryParameters: { project: projectId } },
+      field.id
+    ).observable;
+
+    const subscription = observable.subscribe(
+      (usersByField: TStreamResponse | NilOrError) => {
+        if (isNilOrError(usersByField)) {
+          setReferenceData(usersByField);
+          setIncludedUserPercentage(usersByField);
+          return;
+        }
+
+        if (!usersByField.series.expected_users) {
+          setReferenceDataUploaded(false);
+          return;
+        }
+
+        const referenceData = toReferenceData(usersByField);
+        const includedUsersPercentage = getIncludedUserPercentage(usersByField);
+        setReferenceData(referenceData);
+        setIncludedUserPercentage(includedUsersPercentage);
+        setReferenceDataUploaded(true);
+      }
+    );
+
+    return () => subscription.unsubscribe();
+  }, [code]);
+
+  return {
+    referenceData,
+    includedUserPercentage,
+    referenceDataUploaded,
+  };
+}
+
+export default useReferenceData;
+
+const toReferenceData = (usersByField: TStreamResponse): any => {
+  const { users, expected_users } = usersByField.series;
+  let data = Object.keys(users)
+    .filter((opt) => opt !== '_blank')
+    .map((opt) => ({
+      key: opt,
+      actualPercentage: Math.round(
+        (users[opt] / Object.values(users).reduce((acc, v) => v + acc, 0)) * 100
```

Comment:
You're now running the entire `reduce` loop on every iteration of the `map`. Better to run it only once before line 100, e.g.

```
const totalActualUsers = Object.values(users).reduce((acc, v) => v + acc, 0)
const totalReferenceUsers = Object.values(expected_users).reduce((acc, v) => v + acc, 0)
...
actualPercentage: Math.round((users[opt] / totalActualUsers) * 100),
referencePercentage: Math.round((expected_users[opt] / totalReferenceUsers) * 100)
```

By the way, I am not a huge fan of this terminology coming from the backend. Why `expected_users` for the reference data? They're not expected (unless we expect every single person living in a city to use the platform), and they're not users (they're just inhabitants of the city). Would be nice if we just call it `reference_population` or something like that. Maybe something we can discuss with Adrien

Reply:
cool thanks, should have take a second look to optimize after making it work.

Reply:
ping to @adessy to discuss terminology later, but i think what you said make sense. maybe this will clear out at the discussion being made about how to call the feature.
also representativeness seems too long maybe we can refer to it as r16s (copying a11y or i18n)..